paprika/analyze.py

```diff
@@ -18,69 +17,112 @@ class Analyze(object):
-            beta_restraint = None
-            theta_restraint = None
-
-            for restraint in self.restraints:
-
-                mask2_residue_name = structure[restraint.mask2].residues[0].name
-
-                if "DM" in restraint.mask1 and self.guest_residue_name.upper() in mask2_residue_name and not restraint.mask3 and \
-                        not restraint.mask4:
-                    r_restraint = restraint
-                if restraint.mask3 and not restraint.mask4:
-                    mask3_residue_name = structure[restraint.mask3].residues[0].name
-                    if "DM" in restraint.mask1 and "DM" in restraint.mask2 and self.guest_residue_name.upper() in mask3_residue_name:
-                        theta_restraint = restraint
-                    if "DM" in restraint.mask1 and self.guest_residue_name.upper() in mask2_residue_name and self.guest_residue_name.upper() in \
-                            mask3_residue_name:
-                        beta_restraint = restraint
-            if r_restraint and theta_restraint and beta_restraint:
-                analysis.compute_ref_state_work(
-                    [r_restraint, theta_restraint, None, None, beta_restraint, None]
-                )
-            else:
-                logging.warning("Could not determine the r, θ, or β restraint for computing the analytic release step.")
-
-        else:
-            analysis.compute_free_energy(phases=["release"])
-
-        return analysis
+        analysis.temperature = temperature
+
+        distance_restraint = next(
+            (
+                restraint
+                for restraint in guest_restraints
+                if "DM" in restraint.mask1
+                and restraint.mask2 is not None
+                and restraint.mask3 is None
+                and restraint.mask4 is None
+            ),
+            None
+        )
+
+        theta_restraint = next(
+            (
+                restraint
+                for restraint in guest_restraints
+                if "DM" in restraint.mask1
+                and "DM" in restraint.mask2
+                and restraint.mask3 is not None
+                and restraint.mask4 is None
+            ),
+            None
+        )
+        beta_restraint = next(
+            (
+                restraint
+                for restraint in guest_restraints
+                if "DM" in restraint.mask1
+                and "DM" not in restraint.mask2
+                and restraint.mask3 is not None
+                and restraint.mask4 is None
+            ),
+            None
+        )
+
+        if not distance_restraint or not theta_restraint or not beta_restraint:
+
+            raise RuntimeError(
+                "Could not determine the r, θ, or β restraint for computing the "
+                "analytic release step."
+            )
+
+        analysis.compute_ref_state_work(
+            [distance_restraint, theta_restraint, None, None, beta_restraint, None]
```

Comment:
This could be generalized to include the φ, α, and γ restraints as well.

Reply:
Good point. Probably put this in a separate PR, unless you think it'd be easy to add here. I don't know what this would look like off the top of my head.

Reply:
Yeah, maybe we can do this in a separate PR. FYI the other three restraints are dihedral angles:
φ: D3, D2, D1, G1
α: D2, D1, G1, G2
γ: D1, G1, G2, G3

Reply:
This would be a nice feature, but likely implemented and tested in a separate PR.